src/views/preview/project-view.jsx

```diff
@@ -219,6 +223,84 @@ class Preview extends React.Component {
+
+    // This is copy of what is in save-project-to-server in GUI that adds
+    // an extra get of the project info from api.  We  do this to wait for replication
+    // lag to pass.  This is intended to be a temporary fix until we use the data
+    // from the create request to fill the projectInfo state.
+    handleUpdateProjectData (projectId, vmState, params) {
+        const opts = {
+            body: vmState,
+            // If we set json:true then the body is double-stringified, so don't
+            headers: {
+                'Content-Type': 'application/json'
+            },
+            withCredentials: true
+        };
+        const creatingProject = projectId === null || typeof projectId === 'undefined';
+        const queryParams = {};
+        if (params.hasOwnProperty('originalId')) queryParams.original_id = params.originalId;
+        if (params.hasOwnProperty('isCopy')) queryParams.is_copy = params.isCopy;
+        if (params.hasOwnProperty('isRemix')) queryParams.is_remix = params.isRemix;
+        if (params.hasOwnProperty('title')) queryParams.title = params.title;
+        let qs = queryString.stringify(queryParams);
+        if (qs) qs = `?${qs}`;
+        if (creatingProject) {
+            Object.assign(opts, {
+                method: 'post',
+                url: `${this.props.projectHost}/${qs}`
+            });
+        } else {
+            Object.assign(opts, {
+                method: 'put',
+                url: `${this.props.projectHost}/${projectId}${qs}`
+            });
+        }
+        return new Promise((resolve, reject) => {
+            xhr(opts, (err, response) => {
+                if (err) return reject(err);
+                if (response.statusCode !== 200) return reject(response.statusCode);
+                let body;
+                try {
+                    // Since we didn't set json: true, we have to parse manually
+                    body = JSON.parse(response.body);
+                } catch (e) {
+                    return reject(e);
+                }
+                body.id = projectId;
+                if (creatingProject) {
+                    body.id = body['content-name'];
+                }
+                resolve(body);
+            });
+        }).then(body => {
+            const fetchProjectInfo = (count, resolve) => {
+                api({
+                    uri: `/projects/${body.id}`,
+                    authentication: this.props.user.token
+                }, (err, projectInfo, response) => {
+                    if (err) {
+                        log.error(`Could not fetch project after creating: ${err}`);
+                        return resolve(body);
+                    }
+                    if (typeof body === 'undefined' || response.statusCode === 404) {
+                        // Retry after 500ms, 1.5s, 3.5s, 7.5s and then stop.
+                        if (count > 4) {
+                            return resolve(body);
+                        }
```

Comment:
What does the line below do? Does it really return what we want it to return?

Reply:
Technically it is returning the id from setTimeout, but we don't actually care about the return value. We want it to stop the function execution. What really matters is the resolve a few lines below.